### PR TITLE
File to s3 draining

### PR DESCRIPTION
**Pull request by SimonRichardson**
The changeset adds draining from file-backed storage to S3-backed storage, without creating a new controller and migrating the model. By starting the controller as file-backed storage, you can enable S3 using minio or alternative S3 implementation and then `juju enable-ha` to provide a production environment. This co-exists with just going straight to HA while using S3.

The concept of draining from file-backed to S3-backed is only available to the bootstrap controller. Thus attempting to drain a non-bootstrap controller is ignored. 

Draining happens per object-store worker, so each model object-store handles its own draining. It will do this one file at a time, and that's interleaved with make requests. If the object is not located in the S3-backed storage, it will query the local file-based system. This will only happen if it's draining and if the file exists. Once each file is drained to S3, each file is removed. The S3-backed storage will become the source of truth. Currently, there is no way to know if a model has drained all of its files, it will repeat the process upon every restart to ensure that all files are drained. A future PR will be required to add that concept in (discussion required).

In addition to the draining code, the following changes are also implemented:

 - The agent config now stores the object-store-type, so it can handle restarts correctly.
 - Controller config from `file` to `s3` is blocked unless all s3 config values are correct (we could add this check for bootstrap as well, to prevent bricked bootstraps out of the box).
 - We provide on-the-fly rehashing from sha384 to sha256, whilst optimising for `io.ReadSeeker` interfaces (we don't need additional memory in those cases).

## Checklist

- [x] Code style: imports ordered, good names, simple structure, etc
- [x] Comments saying why design decisions were made
- [x] Go unit tests, with comments saying what you're testing

#### S3 based

We're going to use local s3 setup, but this works just as well with AWS S3.

Install docker or podman (I'll demonstrate using docker).

```sh
docker run -d \
   -p 9000:9000 \
   -p 9001:9001 \
   --user $(id -u):$(id -g) \
   --name minio1 \
   -e "MINIO_ROOT_USER=ROOTUSER" \
   -e "MINIO_ROOT_PASSWORD=CHANGEME123" \
   -v ${HOME}/.local/minio/data:/data \
   quay.io/minio/minio server /data --console-address ":9001"
```

Go to http://localhost:9001/access-keys and add an access key, keeping note of the key and secret.

Locate the IP for minio, we'll use that later.

```sh
$ juju bootstrap lxd test --build-agent --config="object-store-type=s3" --config="object-store-s3-endpoint=http://${IP_ADDRESS}:9000" --config="object-store-s3-static-key=${S3_KEY}" --config="object-store-s3-static-secret=${S3_SECRET}"
```

Once deployed you can inspect minio bucket.

#### File based

```sh
$ juju bootstrap lxd test --build-agent
```

### Drain tests:

This will bootstrap to file based storage by default:

```sh
$ juju bootstrap lxd test --build-agent
```

The following will fail because a valid s3 config is required before moving to s3 backed storage:

```sh
$ juju controller-config "object-store-type=s3"
$ juju controller-config "object-store-type=s3" "object-store-s3-endpoint=http://192.168.0.60:9000" "object-store-s3-static-key=<SECRET_KEY>"
```

This allows you to fill in the full config and at the right time, switch the object-store-type over. You can not go back though. Once you're on s3, you're there to stay.

This swaps it over in one command.

```sh
$ juju controller-config "object-store-type=s3" "object-store-s3-endpoint=http://192.168.0.60:9000" "object-store-s3-static-key=<SECRET_KEY>" "object-store-s3-static-secret=<SECRET_TOKEN>"
```

You can then check the logs to see the draining process.

Notice that the files should have been removed from the file storage once everything has been drained.

### Test flow

```sh
$ juju enable-ha
$ juju add-model default
$ juju deploy juju-qa-test --resource foo-file="./tests/suites/resources/foo-file.txt"
$ juju deploy ubuntu -n 3
$ juju destroy-model default
```

## Documentation changes

@tmihoc We should document this flow, to ensure that people are aware that by default just bootstrapping juju isn't not production grade. Only moving to S3 is classed as production-ready, which allows HA enablement.

## Links

**Jira card:** JUJU-5235

### CODE REVIEW

File: controller/config.go [hunk 1]
```diff
@@ -1470,3 +1474,17 @@ func parseRatio(c Config, name string) (float64, error) {
 		return 0, errors.Errorf("unexpected type %T", c[name])
 	}
 }
+
+// HasCompleteS3Config returns true if the controller has a complete S3
```
Comment (manadart) [inline controller/config.go:1478]: Wrong comment.

File: controller/config.go [hunk 2]
```diff
@@ -1470,3 +1474,17 @@ func parseRatio(c Config, name string) (float64, error) {
...
+// configuration. This includes an endpoint, static key, and static secret.
+func HasCompleteS3ControllerConfig(cfg Config) bool {
```
Comment (manadart) [inline controller/config.go:1480]: Why not a member?
Edit: nevermind; I see it now.

File: core/objectstore/objectstore.go [hunk 3]
```diff
@@ -38,6 +38,9 @@ type Session interface {
 
 // ReadSession provides read access to the object store.
 type ReadSession interface {
+	// HeadObject returns nil if the object exists, or an error if it does not.
+	HeadObject(ctx context.Context, bucketName, objectName string) error
```
Comment (manadart) [inline core/objectstore/objectstore.go:42]: `CheckExists` or `ValidateExists`?

File: domain/controllerconfig/service/service.go [hunk 4]
```diff
@@ -174,15 +174,30 @@ func validObjectStoreProgression(current map[string]string, updateAttrs controll
...
 	// filestorage to s3storage.
-	if cur == upd || cur == objectstore.FileBackend && upd == objectstore.S3Backend {
+	if cur == upd {
 		return nil
+	} else if cur == objectstore.FileBackend && upd == objectstore.S3Backend {
+		// To be 100% sure that we can change from filestorage to s3storage,
+		// we're going to check if the current config has a complete s3 config.
+		// This is rather expensive, but it's the only way to be sure that we
+		// can change from filestorage to s3storage.
+		if controller.HasCompleteS3ControllerConfig(updateAttrs) || hashCompleteS3Config(current) {
```
Comment (manadart) [inline domain/controllerconfig/service/service.go:197]: You could make this return nil or an error and give more detail as to what's wrong.

File: domain/controllerconfig/service/service.go [hunk 5]
```diff
@@ -196,6 +211,13 @@ func contains(s []string, e string) bool {
 	return false
 }
 
+func hashCompleteS3Config(config map[string]string) bool {
```
Comment (manadart) [inline domain/controllerconfig/service/service.go:214]: Couldn't you coerce and use the one from the controller config package? Function is misspelled in any case.

File: internal/bootstrap/bootstrap.go [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+)
+
+// BootstrapParamsPath returns the path to the bootstrap params file.
+func BootstrapParamsPath(dataDir string) string {
+	return filepath.Join(dataDir, cloudconfig.FileNameBootstrapParams)
+}
+
+// IsBootstrapController returns whether the controller is a bootstrap
+// controller.
+func IsBootstrapController(dataDir string) bool {
```
Comment (manadart) [inline internal/bootstrap/bootstrap.go:20]: I think ID == 0 accomplishes this...

File: internal/objectstore/fileobjectstore.go [hunk 7]
```diff
@@ -206,14 +206,14 @@ func (t *fileObjectStore) Remove(ctx context.Context, path string) error {
 func (t *fileObjectStore) loop() error {
 	// Ensure the namespace directory exists, along with the tmp directory.
 	if err := t.ensureDirectories(); err != nil {
-		return errors.Trace(err)
+		return errors.Annotatef(err, "ensure directory")
```
Comment (manadart) [inline internal/objectstore/fileobjectstore.go:209]: "ensuring file store directories exist"

File: internal/objectstore/hash.go [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+	return &hashFileSystemAccessor{
+		fs:        os.DirFS(path),
+		path:      path,
+		namespace: namespace,
+		logger:    logger,
+	}
+}
+
+// HeadHash checks if the file at hash exists in the file storage.
+func (t *hashFileSystemAccessor) HeadHash(ctx context.Context, hash string) error {
```
Comment (manadart) [inline internal/objectstore/hash.go:34]: Same as before; something explicit `ValidateFileAndHashExist`.

File: internal/objectstore/s3objectstore.go [hunk 9]
```diff
@@ -205,23 +266,26 @@ func (t *s3ObjectStore) Remove(ctx context.Context, path string) error {
...
 }
 
 func (t *s3ObjectStore) loop() error {
+	// Report the initial started state.
+	t.reportInternalState(stateStarted)
+
 	// Ensure the namespace directory exists, along with the tmp directory.
 	if err := t.ensureDirectories(); err != nil {
-		return errors.Trace(err)
+		return errors.Annotatef(err, "ensure directory")
```
Comment (manadart) [inline internal/objectstore/s3objectstore.go:281]: Same as before - more text.

File: internal/objectstore/s3objectstore.go [hunk 10]
```diff
@@ -26,6 +27,22 @@ const (
 	defaultPruneInterval = time.Hour * 6
 )
 
+// FileSystemAccessor is the interface for reading and deleting files from the
```
Comment (manadart) [inline internal/objectstore/s3objectstore.go:30]: Wrong comment.

File: internal/objectstore/s3objectstore.go [hunk 11]
```diff
@@ -315,8 +419,29 @@ func (t *s3ObjectStore) get(ctx context.Context, path string) (io.ReadCloser, in
...
+	}); err != nil && !errors.Is(err, errors.NotFound) {
+		return nil, -1, errors.Annotatef(err, "get object")
+	} else if errors.Is(err, errors.NotFound) {
+		// If we're not allowed to use the file accessor, then we can't
+		// attempt to get the file from the file backed object store.
+		if useAccessor == noFileFallback {
+			return nil, -1, errors.Trace(err)
+		}
+
+		// If we're allowed to use the file accessor, then we can attempt to
```
Comment (manadart) [inline internal/objectstore/s3objectstore.go:431]: Copy-pasta from above.

File: internal/s3client/client.go [hunk 12]
```diff
@@ -126,6 +128,27 @@ func NewS3Client(endpoint string, httpClient HTTPClient, credentials Credentials
 	}, nil
 }
 
+// HeadObject checks if an object exists in the object store based on the bucket
+// name and object name.
+// Returns nil if the object exists, or an error if it does not.
+func (c *S3Client) HeadObject(ctx context.Context, bucketName, objectName string) error {
```
Comment (manadart) [inline internal/s3client/client.go:134]: That opaque name again.

**Review by manadart — COMMENTED** [review]
Code is mostly good. Some suggestions. Will QA first thing.

File: internal/bootstrap/bootstrap.go (hunk 6 above)
Comment (SimonRichardson) [reply internal/bootstrap/bootstrap.go:20]: It does, but having a bootstrapped params also guarantees that.

**Review by manadart — APPROVED** [review]
QA works OK, but see my prior comment about more information regarding settings.

For example:
```
11:36:09 github.com/juju/juju (HEAD detached at simonrichardson/file-to-s3-draining)-> juju controller-config "object-store-type=s3"
ERROR updating controller config state: can not change "object-store-type" from "file" to "s3" without complete s3 config
11:36:19 github.com/juju/juju (HEAD detached at simonrichardson/file-to-s3-draining)-> juju controller-config object-store-s3-static-secret
ERROR key "object-store-s3-static-secret" not found in controller "lxd"
11:37:17 github.com/juju/juju (HEAD detached at simonrichardson/file-to-s3-draining)-> juju controller-config "object-store-s3-endpoint=http://192.168.5.101:9000"
11:37:46 github.com/juju/juju (HEAD detached at simonrichardson/file-to-s3-draining)-> juju controller-config "object-store-s3-endpoint"
http://192.168.5.101:9000
11:37:52 github.com/juju/juju (HEAD detached at simonrichardson/file-to-s3-draining)-> juju controller-config "object-store-s3-static-key=ZG1ShAeqX0M0oHMd6fnx"
11:38:16 github.com/juju/juju (HEAD detached at simonrichardson/file-to-s3-draining)-> juju controller-config "object-store-type=s3"
ERROR updating controller config state: can not change "object-store-type" from "file" to "s3" without complete s3 config
11:38:20 github.com/juju/juju (HEAD detached at simonrichardson/file-to-s3-draining)-> juju controller-config "object-store-s3-static-secret=HVcM3xZHFZCfM9WTtaSXN2WFPd8J0vpHbRToxqSZ"
11:38:54 github.com/juju/juju (HEAD detached at simonrichardson/file-to-s3-draining)-> juju controller-config "object-store-type=s3"
```
I think at each of these statements, we could inform the user of the requirements that are unmet.

Happy to have a follow-up.

**Pull request merged**